File: solutions/Python/kth_smallest_element_in_bst_inorder.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def kthSmallest_inorder(root: Optional[TreeNode], k: int) -> int:
    """
    In-order DFS approach: traverse BST in sorted order, return k-th element.
    O(n) time (worst case), O(h) space.
    """
    result = {"count": 0, "value": None}

    def inorder(node):
        if not node or result["value"] is not None:
            return

        inorder(node.left)

        result["count"] += 1
        if result["count"] == k:
            result["value"] = node.val
            return

        inorder(node.right)

    inorder(root)
    return result["value"]


# Test
root = TreeNode(3)
root.left = TreeNode(1)
root.right = TreeNode(4)
root.left.right = TreeNode(2)

print(kthSmallest_inorder(root, 1))  # 1
print(kthSmallest_inorder(root, 3))  # 2

File: solutions/JavaScript/kth_smallest_element_in_bst_inorder.js
class TreeNode {
    constructor(val = 0, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

function kthSmallest_inorder(root, k) {
    let count = 0;
    let result = null;

    function inorder(node) {
        if (!node || result !== null) return;

        inorder(node.left);

        count++;
        if (count === k) {
            result = node.val;
            return;
        }

        inorder(node.right);
    }

    inorder(root);
    return result;
}

// Test
const root = new TreeNode(3);
root.left = new TreeNode(1);
root.right = new TreeNode(4);
root.left.right = new TreeNode(2);

console.log(kthSmallest_inorder(root, 1));  // 1
console.log(kthSmallest_inorder(root, 3));  // 2
